Database objects periodic import

Tener opened the pull request:
This PR adds periodic import of database objects from supported databases.

Currently implemented for Postgres, which necessitated some refactoring.

Updates https://github.com/gravitational/teleport/issues/32627

Changelog: Automatically import database objects from supported databases per configured import rules.

Tener committed:
Add periodic import of database objects from supported databases.

A single implementation is included: for Postgres, along with necessary changes.

Also refactors SQLAdminClient interface to accept cert expiry instead of pulling that information from `tlsca.Identity`.

Tener committed:
fix logging, imports

Tener committed:
add missing license

greedy52 commented on lib/srv/db/postgres/objects.go:
```diff
@@ -0,0 +1,130 @@
+var _ objects.ObjectFetcher = (*objectFetcher)(nil)
+
+func NewObjectFetcher(ctx context.Context, db types.Database, cfg objects.ObjectFetcherConfig) (objects.ObjectFetcher, error) {
+	if db.GetAdminUser().DefaultDatabase == "" || db.GetAdminUser().Name == "" {
```
hmm. `default_database` is not really required for Postgres auto-user provisioning, and not used at all for the current DAC implementation IIRC.

Maybe pick a database name from the non-wildcard rules? Or default to "postgres" database?

Tener replied:
Hmm, I'm not certain how safe that heuristic is, but I like the idea of reducing the configuration requirements. What about trying either `postgres` or `teleport-admin`?

greedy52 replied:
Let's do `postgres` so users don't have to create a new database. we can make the doc clear that either the user gives admin user permission to connect `postgres` or specify a default database.

Tener replied:
Makes sense, I'll do so.

Tener replied:
Done: the requirement for db name was dropped, we default to "postgres" in "get all db names" function.

greedy52 commented on lib/srv/db/postgres/connector.go:
```diff
@@ -0,0 +1,166 @@
+	startupParams map[string]string
+}
+
+func (c *connector) copy() *connector {
+	return &connector{
+		auth:          c.auth,
+		cloudClients:  c.cloudClients,
+		certExpiry:    c.certExpiry,
+		database:      c.database,
+		databaseUser:  c.databaseUser,
+		databaseName:  c.databaseName,
+		startupParams: c.startupParams,
+	}
+}
```
why not copy like this:
https://github.com/gravitational/teleport/blob/0f4eadd8d0a34eb6119e58bc2d9cb3bfc3e87b4b/lib/srv/db/common/session.go#L87-L88
Are we explicitly skipping `log`?

Tener replied:
Ah, yes that is much better. Harder to forget about fields in this way 😅

Tener commented on lib/srv/db/objects/importer.go:
```diff
@@ -0,0 +1,249 @@
+		return &noopImporter{}, nil
+	}
+
+	imp := startSingleDatabaseImporter(ctx, cfg, database, fetcher)
+	imp.start(ctx)
```
> currently imp.start is called twice so maybe change startSingleDatabaseImporter to a constructor.
Oops, that should not be the case. I'll fix. I actually had `newSingleDatabaseImporter`, which is why that call is there - a leftover - but it made passing the cancel context somewhat fragile, which is why I went with this solution.

Tener replied:
Refactored to `go imp.start(ctx)` as well as bunch of other related changes.

Tener commented on lib/srv/db/objects/importer.go:
```diff
@@ -0,0 +1,249 @@
+	return imp
+}
+
+func (i *singleDatabaseImporter) stop() {
+	i.cancel()
+}
```
No, as there may be other importers running, the service may only be down temporarily etc. In general I believe it is better to keep these objects around rather than delete them unnecessarily.

greedy52 replied:
sounds good. Let's add a comment on that. i guess that could be part of the future performance improvement.

That said, should `s.cfg.DatabaseObjects.StopImporter` be called in `s.stopProxyingDatabase`? For example, if the `server.Shudown` is called, the importer is not stopped with current impl. IIRC `server.Shudown` can be called during a CA rotation or a systemd reload too.

Tener replied:
Wouldn't the server shutdown cancel the parent context? I'll check out `stopProxyingDatabase` too.

greedy52 replied:
> Wouldn't the server shutdown cancel the parent context? I'll check out `stopProxyingDatabase` too.

good point. as long as it gets stopped during ShutDown, we are good.

Tener commented on lib/srv/db/objects/importer.go:
```diff
@@ -0,0 +1,249 @@
+
+// objWithExpiry holds an object separate to its expiry time.
+type objWithExpiry struct {
+	obj    *dbobjectv1.DatabaseObject
```
That is part of it, sure. I'll add a comment.

Tener commented on lib/srv/db/server.go:
```diff
@@ -82,6 +83,8 @@ func init() {
 	common.RegisterEngine(clickhouse.NewEngine, defaults.ProtocolClickHouseHTTP)
 	common.RegisterEngine(spanner.NewEngine, defaults.ProtocolSpanner)
+
+	objects.RegisterObjectFetcher(postgres.NewObjectFetcher, defaults.ProtocolPostgres)
```
This is done on demand in the unit test: `TestStartDatabaseImporter`

Tener commented on lib/srv/db/objects/importer.go:
```diff
@@ -0,0 +1,249 @@
+			continue
+		}
+
+		// do we need to refresh the object in backend?
+		if objOld.expiry.Sub(now) < cfg.RefreshThreshold {
+			countRefresh++
+			updated[key] = &objWithExpiry{
+				obj:    objNew,
+				expiry: expiry,
+			}
+		}
```
The expiry check is part of the reason: we could pass custom comparer that ignores expiry and it wouldn't send any updates if only the expiry is unchanged.

But we would need to track the expiry in some way anyway and I didn't see a satisfactory way forward.

Overall, I think `services.Reconciler` isn't quite suited for this update strategy and implementing the core logic from scratch leads to better readability. We might make the code generic if in the future if there will be more suitable use cases for it.

Tener commented on lib/srv/db/objects/objects.go:
```diff
@@ -0,0 +1,144 @@
+		c.Clock = clockwork.NewRealClock()
+	}
+
+	if c.ScanInterval == 0 {
+		c.ScanInterval = time.Minute * 5
+	}
+	if c.ObjectTTL == 0 {
+		c.ObjectTTL = time.Minute * 60
+	}
+	if c.RefreshThreshold == 0 {
+		c.RefreshThreshold = time.Minute * 15
+	}
```
Agreed, I'll add.

Tener replied:
Added, including the option to disable the service entirely.

Tener commented on lib/srv/db/objects/fetcher.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+// getObjectFetcher returns a new object fetcher for given database.
+// The returned fetcher may be nil, which indicates that particular database configuration is not supported.
```
Hmm, in general I'm not a fan of tying the control flow to the error values, but I'll see if this can be improved.

greedy52 replied:
> Hmm, in general I'm not a fan of tying the control flow to the error values, but I'll see if this can be improved.

me neither. but I prefer that over returning `nil, nil`. 

Or instead of `trace.NotImplemented`, do a precheck before calling this and return a hard `trace.NotFound` and don't handle it? Or maybe add a bool in this return or something.

Tener replied:
We can check if given protocol is implemented, sure, but the second case for nil fetcher is a particular database configuration that may not support fetching objects.  These are very similar in the end.

Tener replied:
I've simplified this a lot; basically, we are down to standard error handling with the custom bit being lowering the severity of log in case the database is of unsupported protocol.

Tener commented on lib/srv/db/objects/importer.go:
```diff
@@ -0,0 +1,249 @@
+}
+
+func (i *singleDatabaseImporter) fetchObjects(ctx context.Context) ([]*dbobjectv1.DatabaseObject, error) {
+	rules, err := i.cfg.AuthClient.GetDatabaseObjectImportRules(ctx)
+	if err != nil {
+		i.cfg.Log.ErrorContext(ctx, "Failed to fetch database object import rules", "err", err)
+		return nil, trace.Wrap(err)
+	}
+
+	dbNameFilter := databaseobjectimportrule.CalculateDatabaseNameFilter(rules, i.database)
+	objsFetched, err := i.fetcher.FetchDatabaseObjects(ctx, dbNameFilter)
+	if err != nil {
+		i.cfg.Log.WarnContext(ctx, "Error while fetching database objects", "err", err, "imported", len(objsFetched))
+	}
+
+	// TODO(Tener): switch to cfg.log logger once ApplyDatabaseObjectImportRules is on slog too.
+	objsImported, errCount := databaseobjectimportrule.ApplyDatabaseObjectImportRules(logrus.WithContext(ctx).WithField("db", i.database), rules, i.database, objsFetched)
+	counts, _ := permissions.CountObjectKinds(objsImported)
+	i.cfg.Log.InfoContext(ctx, "Database objects imported", "counts", counts, "err_count", errCount, "total", len(objsImported))
```
Looking at this with fresh eyes I think some refactoring is due. This code piece is lifted from Postgres DAC implementation, but it would be good to abstract away this pattern (i.e. fetch rules, fetch objects, apply rules, return finished objects).

Tener replied:
I've done the refactoring I've mentioned above. All object-fetching code is using the same interface now, with import rules being applied consistently by a wrapping fetcher.

Tener committed:
review feedback
- several minor changes
- apply import rules for all registered object fetchers using wrapper type
- refactor DAC to use object fetcher
- adjust tests

Tener committed:
add comment and missing log call

Tener committed:
support config overrides through unstable env vars, including disabling the service entirely

greedy52 commented on lib/srv/db/postgres/users.go:
```diff
@@ -205,33 +198,26 @@ func (e *Engine) applyPermissions(ctx context.Context, sessionCtx *common.Sessio
-	}
-
-	conn, err := e.connectAsAdmin(ctx, sessionCtx, false)
+	fetcher, err := objects.GetObjectFetcher(ctx, sessionCtx.Database, objects.ObjectFetcherConfig{
```
I really like reusing the same fetcher for user sessions. Is it possible to reuse the same admin connection for calling the permission procedures though?

Tener replied:
Reuse sounds tricky to do properly. I'll look into it, I do have an idea that might work. Even if not, perhaps this is acceptable for the ease of maintenance this otherwise brings.

greedy52 replied:
could we cache it in the `fetcher`? something like

```
type cachedAdminPGXConnGetter interface {
  getCachedPGXAdminConn() (*pgx.Conn, error)
}

if cached, ok := fetcher.(cachedAdminPGXConnGetterr); ok {
  conn, err = cached.getCachedAdminConn()
} else {
  conn, err = e.connectAsAdminSessionDatabase(ctx, sessionCtx)
} 
```
Or maybe move the apply part to the fetcher interface as well. I will leave it up to you if you want to keep it like this or we can tackle this later when doing performance improvement.

Tener replied:
Yep, I've tried something along these lines. Trouble is with ensuring the connection params are the same, which often isn't the case. I'll give it another look tomorrow though, perhaps caching the fetcher is indeed the way forward.

greedy52 replied:
>  I'll give it another look tomorrow though, perhaps caching the fetcher is indeed the way forward.

I just gave it another thought that it won't work well as we have to deal with closing the connection.

Tener replied:
Yes, that is another complication I've run into. I've considered using a connection pool instead, but those are not designed for our use case at all.

Tener committed:
refactoring and minor changes

- be more specific about required API for ObjectFetcherConfig
- measure scan time
- bump default scan interval
- call StopImporter from stopDatabase to fix issue with dynamic db resource updates
- add comments

Tener committed:
Refactor `connectAsAdmin`, add `withDefaultDatabase()` method for clarity of implementation.

Tener committed:
Reduce surface area for fetcher config

Tener committed:
make the not-implemented case less noisy by skipping the error

greedy52 commented on lib/srv/db/server.go:
```diff
@@ -465,13 +482,26 @@ func (s *Server) startDatabase(ctx context.Context, database types.Database) err
+		if trace.IsNotImplemented(err) {
+			s.log.DebugContext(ctx, "Database object importer not implemented.", "database", database.GetName())
+		} else {
+			s.log.WarnContext(ctx, "Failed to start database object importer.", "database", database.GetName(), "error", err)
```
Are we still getting this when postgres databases are not configured with auto-user provisioning?

Tener replied:
No, implementation-wise `NotImplemented` is only returned for unsupported protocols.

greedy52 replied:
> No, implementation-wise `NotImplemented` is only returned for unsupported protocols.

a postgres database without admin user will return BadParameter and come here?

Tener replied:
Yes, this will give a warning. Do you think we should handle that case separately?

greedy52 replied:
i mean not setting admin user is a common case, not an error case. we shouldn't error/warn in that case, or maybe we shouldn't even start importer when admin user is not set.

Tener replied:
I took it from the perspective of answering the question of "why does the importer doesn't work for my database", but you are right, this doesn't deserve a treatment of an error. I'll add handling for this.

Tener replied:
We will now get this sort of message (debug verbosity):
```
2024-07-12T15:20:30+02:00 DEBU [DB:SERVIC] Database object importer cannot be started due to disabled fetcher reason:no admin user configured database:mypg db/server.go:491
```

Tener commented on lib/srv/db/objects/objects.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+type Config struct {
+	AuthClient   *authclient.Client
```
Done.

Tener commented:
@timothyb89 do you mind giving it a look?

Tener committed:
lint fix

Tener committed:
decrease verbosity of reporting for unsupported configurations

GavinFrazar commented on lib/srv/db/objects/objects.go:
```diff
@@ -0,0 +1,208 @@
+	// overriding scan interval modifies the other variables, but not the other way around.
+	if found, value := c.loadEnvVar(ctx, "TELEPORT_UNSTABLE_DB_OBJECTS_SCAN_INTERVAL"); found {
+		c.ScanInterval = value
+		c.ObjectTTL = value * 12
+		c.RefreshThreshold = value * 3
```
could you add a comment explaining why the numbers 12 and 3?
It may be a good idea also to add the unit of time to these env vars to make it clear, e.g. `TELEPORT_UNSTABLE_DB_OBJECTS_OBJECT_TTL_MINUTES` (I'm actually not sure what the units are assumed to be currently)

edit: ok I see `time.ParseDuration` handles a wide range of time formats, including negative values.
I think I still prefer something like `_MINUTES` and just parsing an unsigned int, with a check for the 0 value if 0 is invalid. That enforces a minimum value of 1 minute automatically, and is simpler to reason about plus the env var name documents itself.

Tener replied:
I think I picked this pattern from elsewhere in Teleport, feels good to keep to one convention. The updated logic checks for zero/negative intervals which matter, so we should be good. In any case messing with these `UNSTABLE` env vars puts responsibility on the administrator to pick sensible values and monitor the stability afterwards.

Tener committed:
make `reason` field private

Tener committed:
move `unregisterObjectFetcher` to tests

Tener committed:
avoid exceedingly long lines with log calls

Tener committed:
correct formatting leftover from slog conversion

Tener committed:
tighten logic around disabled importer

if the scan interval is zero or less, the importer becomes disabled, which avoids panic from interval.New().

add some tests and comments for clarity.

Tener commented on lib/srv/db/objects/objects.go:
```diff
@@ -0,0 +1,208 @@
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	if c.ScanInterval == 0 {
+		c.ScanInterval = time.Minute * 15
+	}
+	if c.ObjectTTL == 0 {
+		c.ObjectTTL = time.Minute * 180
+	}
+	if c.RefreshThreshold == 0 {
+		c.RefreshThreshold = time.Minute * 45
+	}
+
+	c.loadEnvVarOverrides(ctx)
```
I'm not sure about minimum values. In general, the env var overrides are there for very low-level access. Good point about avoiding panics, however. I've modified the `disabled()` method to check for zero or negative interval.

Tener committed:
Merge branch 'master' into tener/dac-import

Tener committed:
Merge branch 'master' into tener/dac-import

Tener commented:
@r0mant this PR needs admin approval, can you provide one?

Tener committed:
drop unused method

r0mant reviewed (approved):
Bot.
